### ISSUE TITLE
Document Scope of Custom Tool Parameters

### ISSUE THREAD

**Issue by jmchilton** [galaxyproject/galaxy#3413]
There has been some push back (in various pull requests and issues and to varying degrees) related to how configurable tool parameters should be. I'd suggest we as a community discuss these in one place and then document somehow our decisions (not that anything is final - the scope of Galaxy constantly changes I think as the use cases driving development do). It would merely serve to focus future efforts and be a metric by which we can evaluate pull requests a bit more consistently using.

- What level of generality do we want to insist on for tool parameters? Something like ``data`` or ``text`` seems very generic, ``color`` a little less so but still broadly applicable. Something like ``genomebuild`` is only useful to sequencing applications.;
  - Would we accept ``genomebuild`` today?
  - Would we accept something comparable from another field we target - such as ``language`` for NLP?
  - Would we accept something more specific but relate-able to a datatype stored in Galaxy core.
  - Will we accept a password parameter #3121?
  - We we accept a PR that adds the ability to upload a file directly to a tool.
- For parameter types not in Galaxy - what level of extension stability do we want to provide?
  - Do we support a generic framework with a stable API required code modifications?
  - Would we support a plugin framework such as #2751 for these things - and if so what would that need to look like?
- What degree of customization should we allow for workflows?
  - Should we support convenience features such as connecting data sources or uploads or collection creation directly to workflow execution?
  - Should we support a framework for that?
  - Would we accept a PR that adds the ability to upload a file directly to a workflow execution.

**Comment by jxtx** [comment]
> Would we accept genomebuild today?

I'll start with this one. Ideally this would just be a "select" type parameter with the possible values taken from a controlled vocabulary. Galaxy should support registering different such vocabularies and tools should be able to reference then. Datatype metadata fields as well, e.g. dbkey should also be runtime configured and would share the same vocabulary with genomebuild.

**Comment by bgruening** [comment]
> - What level of generality do we want to insist on for tool parameters? Something like ``data`` or ``text`` seems very generic, ``color`` a little less so but still broadly applicable. Something like ``genomebuild`` is only useful to sequencing applications.;

>  - Would we accept ``genomebuild`` today?

I don't think so. I have already belly ache by `forward`, `reverse` in collections.

>  - Would we accept something comparable from another field we target - such as ``language`` for NLP?

Such things should be configurable or plugin-based if possible. I guess this is what @jxtx means with controlled vocabulary.

>  - Would we accept something more specific but relate-able to a datatype stored in Galaxy core.

datatypes should ideally not be in core. But we do not have a better solution until now - it works and can be more easily tested. I think we should make it easier to strip out datatypes from a flavor.

>  - Will we accept a password parameter #3121?

I don't like this and would like to see Galaxy interacting more with a user-preference store. Which again don't need to be in Galaxy database - it would be confd or similar.

>  - We we accept a PR that adds the ability to upload a file directly to a tool.

Probably, I'm a little bit worried about usability and it needs to be discussed what really happens with this input dataset - essentially it needs to be transparent and not complicated. 

>For parameter types not in Galaxy - what level of extension stability do we want to provide?
>  - Do we support a generic framework with a stable API required code modifications?

I guess this needs to be decided on a case to case basis. Such additions as color etc are great, others might not be. Imho a huge benefit of Galaxy UI is that all tools look to some degree similar and comfortable. A user should not be surprised every time he/she sees a new tool.

>  - Would we support a plugin framework such as #2751 for these things - and if so what would that need to look like?

I hope that most of these usecase can be tackled by enhancing the validators and the filters. Many people don't know them. The IUC is using this to less ... we need to tell people what is possible. You already can filter a select box based on another select box ... this is just not documented well.

> - What degree of customization should we allow for workflows?
>  - Should we support convenience features such as connecting data sources or uploads or collection creation directly to workflow execution?

Connecting data sources is a thing I could imagine to be useful, as long as we store the downloaded raw data as well in the history. The rest I think is currently so complex that a dedicated UI is necessary and it is probably ok like it is today. 
(One thing that could be improved is uploading directly into a collection, and collection support in the library.)
In general we should keep in mind that the traditional Galaxy UI can not fulfill all use-cases and that projects like the Refinery Platform can fill such gaps. I somehow hope that the shift to React will enable us a more flexible and plug-able UI.

>  - Should we support a framework for that?
>  - Would we accept a PR that adds the ability to upload a file directly to a workflow execution.

Same as above for tools.

**Comment by jgoecks** [comment]
I concur with what @jxtx and @bgruening have written. 

That said, IMO many of your questions deserve their own PR/issue for discussion, especially the second and third bullets/subbullets. Do you have in mind that you want to build a list of extensions that would be welcome and those that would unlikely to be welcome? This seems difficult without use cases, implementation details, and end product.
